### PR TITLE
EiffelWeb JWT extension, Oauth JWT token flow, OpenSSL RSA crypto algorithms

### PR DESCRIPTION
**Pull request by jvelilla**

### CODE REVIEW

File: Src/contrib/library/web/authentication/jwt_openssl_ext/src/jwt_alg_rs384.e [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+feature {NONE} -- Implementation		
+
+	sign_rsa384 (s: READABLE_STRING_8; a_secret: READABLE_STRING_8): STRING_8
+		local
+			rs256: SSL_RSA
+			l_priv_key: SSL_RSA_PRIVATE_KEY
+			l_result: STRING_8
+		do
+				--| todo missing error handling.
+			create rs256.make
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/jwt_openssl_ext/src/jwt_alg_rs384.e]: bad local var name .. should be rsa384 or just rsa

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e [hunk 2]
```diff
@@ -0,0 +1,101 @@
+note
+	description: "Summary description for {BOX_OAUTH_20_JWT_API_EXAMPLE}."
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e:2]: replace BOX by Google.

Comment (jocelyn) [reply Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e:2]: It is missing a readme note ... explaining one has to set the attribute values at the bottom.
 And how to get them ...

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			api_service: OAUTH_20_JWT_SERVICE
+			request: OAUTH_REQUEST
+			access_token: OAUTH_TOKEN
+		do
+			create oauth_jwt.make (authentication_url)
+			create config.make (jwt_signing, client_id)
+			api_service := {OAUTH_20_JWT_SERVICE} / oauth_jwt.create_service (config)
+			print ("%N===Box OAuth Server to Server Workflow ===%N")
+
+            if attached api_service then
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e:32]: issue with indentation

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+create
+	make
+
+feature -- Access
+
+	make
+		local
+			oauth_jwt: OAUTH_20_JWT_API
+			config: OAUTH_JWT_CONFIG
+			api_service: OAUTH_20_JWT_SERVICE
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/box_oauth_20_jwt_api_example.e:22]: it is enough to declare as OAUTH_SERVICE_I 
and then later ... remove the "{OAUTH_20_JWT_SERVICE} / ..."

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/examples.ecf [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="ISO-8859-1"?>
-<system xmlns="http://www.eiffel.com/developers/xml/configuration-1-16-0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.eiffel.com/developers/xml/configuration-1-16-0 http://www.eiffel.com/developers/xml/configuration-1-16-0.xsd" name="examples" uuid="533DEB07-C4A6-4EEE-8EC1-4F09DC203A26">
+<system xmlns="http://www.eiffel.com/developers/xml/configuration-1-18-0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.eiffel.com/developers/xml/configuration-1-18-0 http://www.eiffel.com/developers/xml/configuration-1-18-0.xsd" name="examples" uuid="533DEB07-C4A6-4EEE-8EC1-4F09DC203A26">
 	<target name="oauth_examples" abstract="true">
-		<option warning="true" is_obsolete_routine_type="true">
+		<option warning="true" is_obsolete_routine_type="true" manifest_array_type="mismatch_warning">
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/examples.ecf:4]: please remove this `is_obsolete_routine_type="true"` ...!

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/google_oauth_20_jwt_api_example.e [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		do
+			create oauth_jwt.make (authentication_url)
+			create config.make (jwt_signing, client_id)
+			config.set_api_secret (client_secret)
+			api_service := {OAUTH_20_JWT_SERVICE} / oauth_jwt.create_service (config)
+			print ("%N===Box OAuth Server to Server Workflow ===%N")
+
+            if attached api_service then
+					-- client id and client secreat are required by BOX
+            	api_service.enable_client_id
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/examples/google_oauth_20_jwt_api_example.e]: bad indentation

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	revision: "$Revision$"
+
+class
+	OAUTH_JWT_CONFIG
+
+inherit
+
+	OAUTH_CONFIG
+		rename
+			make as make_config
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e:13]: I find suspicious that we inherit from OAUTH_CONFIG ... and then we barely use the ancestor ...

maybe in `make`... call `			make_default (a_api_key, "")`

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		do
+			create api_key.make_from_string (a_api_key)
+			create api_secret.make_empty
+			set_grant_type ("urn:ietf:params:oauth:grant-type:jwt-bearer")
+			create assertion.make_from_string (a_assertion)
+		end
+
+feature -- Access
+
+	assertion: IMMUTABLE_STRING_8
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e:30]: Is this the "standard" name?

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		rename
+			make as make_config
+		end
+
+create
+	make
+
+feature {NONE} -- Implementation
+
+	make (a_assertion: READABLE_STRING_8; a_api_key: READABLE_STRING_8)
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e:21]: I would swap the two arguments, to looks more like ancestor creation proc.

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		end
+
+feature -- Access
+
+	assertion: IMMUTABLE_STRING_8
+			--  JWT bearer token, base64url-encoded.
+
+
+feature -- Change Element
+
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e:34]: remove extra empty line

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+	set_api_secret (a_secret: READABLE_STRING_8)
+		do
+			create api_secret.make_from_string (a_secret)
+		end
+
+;note
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/data/oauth_jwt_config.e]: useless ;

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/oauth/oauth_20_jwt_service.e [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+	is_client_secret_required: BOOLEAN
+			-- Is client secret required?, by default false.
+
+	is_client_id_required: BOOLEAN
+			-- Is client secret required?, by default false.
+
+feature -- Change Element
+
+	enable_client_secret
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/oauth/oauth_20_jwt_service.e:29]: should we also have  disable_client_secret ?

Comment (jocelyn) [reply Src/contrib/library/web/authentication/oauth/cypress/consumer/src/oauth/oauth_20_jwt_service.e:29]: should it be  `require_client_secret` instead?

File: Src/contrib/library/web/authentication/oauth/cypress/consumer/src/oauth/oauth_20_jwt_service.e [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+			-- retrieve an access token using a request token
+			-- (obtained previously)
+		local
+			l_request: OAUTH_REQUEST
+		do
+			create l_request.make (api.access_token_verb, api.access_token_endpoint)
+			if attached config.grant_type as l_grant_type then
+				l_request.add_body_parameter ({OAUTH_CONSTANTS}.grant_type, l_grant_type)
+			end
+			if is_client_id_required and then not config.api_key.is_empty then
```
Comment (jocelyn) [inline Src/contrib/library/web/authentication/oauth/cypress/consumer/src/oauth/oauth_20_jwt_service.e:55]: is_empty is fine, but I tends to use  `is_whitespace` to exclude empty case, and blank text as well.

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_key_pair.e [hunk 14]
```diff
@@ -32,6 +32,7 @@ feature {NONE} -- Initialization
 
 			l_priv_key: C_STRING
 			l_pub_key: C_STRING
+			l_temp: STRING
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_key_pair.e:35]: is this l_temp needed ?

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e [hunk 15]
```diff
@@ -147,13 +147,58 @@ feature -- Access: Public Encrypt - Private Decrypt
...
+			-- Ensures that the signature matches the original code.
+			-- Using the public key `a_pub_key' and the signature encoded in BASE64 with the origin text `a_plain_text'.
+			-- usign the same algorithm as the author of the signature.
+		do
+			Result := verify_implementation (a_pub_key, a_plain_text, a_signature, {SSL_CRYPTO_EXTERNALS}.c_evp_sha512)
+		end
+
+feature {NONE} -- Implementation
+
+	sign_implementation (a_priv_key: SSL_RSA_PRIVATE_KEY; a_message: READABLE_STRING_GENERAL; a_sha: POINTER): detachable STRING_8
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e:198]: I would have named it  `imp_sign` or `implementation_sign`.. or just `sign`...
here it looks ... "sign the implementation.."

same comment for `verify`

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e [hunk 16]
```diff
@@ -202,11 +250,123 @@ feature -- Sign
...
+			l_padding : INTEGER
+
+			l_b64message: C_STRING
+			l_buffer: C_STRING
+			l_ret: INTEGER
+			l_character: CHARACTER_8
+		do
+			Result := True
+
+			l_count := a_signature.count
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e]: the 8 next lines are useless.

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e [hunk 17]
```diff
@@ -202,11 +250,123 @@ feature -- Sign
...
+					-- last two chars are =
+			if a_signature.at (l_count) = '=' and then a_signature.at (l_count - 1) = '=' then
+				l_padding := 2
+			else if a_signature.at (l_count) = '=' then
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e]: avoid the "else if"...   either  "elseif"  or "else %N if "..

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e [hunk 18]
```diff
@@ -202,11 +250,123 @@ feature -- Sign
...
+				l_padding := 1
+			end
+				l_count := (l_count * 3) // 4 - l_padding
+			end
+
+			l_decoded_sig :=(create {BASE64}).decoded_string (a_signature)
+			l_pubkey := {SSL_CRYPTO_EXTERNALS}.c_evp_pkey_new
+			l_res := {SSL_CRYPTO_EXTERNALS}.c_evp_pkey_assign_rsa (l_pubkey, a_pub_key.rsa)
+			if l_res = 0 then
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e:291]: should it be  `l_res <= 0`?

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e [hunk 19]
```diff
@@ -202,11 +250,123 @@ feature -- Sign
...
+				create last_error.make (({SSL_CRYPTO_EXTERNALS}.c_error_get_error))
+				Result := False
+			end
+
+			l_verify_ctx:= {SSL_CRYPTO_EXTERNALS}.c_evp_md_ctx_new
+			l_res := {SSL_CRYPTO_EXTERNALS}.c_evp_digestinit_ex (l_verify_ctx, a_alg, default_pointer)
+			if l_res /= 1 then
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e:298]: should it be  `l_res <= 0`? 

(or change the other  to be `l_res /= 1`)

File: Src/unstable/library/network/openssl/src/crypto/ciphers/wrapper/ssl_crypto_externals.e [hunk 20]
```diff
@@ -451,29 +466,83 @@ feature -- Access
 		end
 
 	c_evp_digestsigninit (a_ctx: POINTER; a_pctx: POINTER; a_type: POINTER; a_engine: POINTER; a_key: POINTER): INTEGER
-			-- --https://www.openssl.org/docs/man1.1.0/crypto/EVP_DigestSignInit.html
+			-- https://www.openssl.org/docs/man1.1.0/crypto/EVP_DigestSignInit.html
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/ciphers/wrapper/ssl_crypto_externals.e:469]: you could also use EIS ... but comment is also fine for now.

File: Src/unstable/library/network/openssl/src/crypto/ciphers/wrapper/ssl_crypto_externals.e [hunk 21]
```diff
@@ -574,14 +658,31 @@ feature -- RSA
...
 			return rsa;
 		  }"
 		end
 
+
+	c_rsa_sign (a_type: INTEGER; a_m: POINTER; a_m_len: INTEGER; a_sigret: POINTER; a_siglen: TYPED_POINTER [INTEGER]; a_rsa:  POINTER): INTEGER
+		external
+			"C inline use %"eif_openssl.h%""
+		alias
+		 " return RSA_sign((int)$a_type, (const unsigned char *)$a_m, (unsigned int)$a_m_len, (unsigned char *)$a_sigret, (unsigned int *)$a_siglen, (RSA *)$a_rsa);"
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/ciphers/wrapper/ssl_crypto_externals.e:676]: bad indentation

File: Src/unstable/library/network/openssl/test/application.e [hunk 22]
```diff
@@ -18,10 +18,11 @@ feature {NONE} -- Initialization
 	make
 			-- Run application.
 		do
-			initialize_ssl
-			test_ecc
-			test_encrypt_decrypt_set_keys
-			test_generate_keys
+--			initialize_ssl
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/test/application.e:21]: is it wanted to have `initialize_ssl` commented?

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e [hunk 23]
```diff
@@ -147,13 +147,58 @@ feature -- Access: Public Encrypt - Private Decrypt
 			create Result.make_from_string (l_decrypt.string)
 		end
 
-
 feature -- Sign
 
-	sign (a_priv_key: SSL_RSA_PRIVATE_KEY; a_message: READABLE_STRING_GENERAL): detachable STRING_8
+	sign_sha256 (a_priv_key: SSL_RSA_PRIVATE_KEY; a_message: READABLE_STRING_GENERAL): detachable STRING_8
```
Comment (jocelyn) [inline Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e:152]: Not critical, but this is a bad name for a function, ... a function should not be named with a verb.

it should be
`sha256_signed_message`  or something like that.

**Review by jocelyn — CHANGES_REQUESTED** [review]
you can pull from https://github.com/jocelyn/EiffelStudio/tree/crypto_jwt_20181105  for part of my comments.

File: Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e (hunk 19 above)
Comment (jvelilla) [reply Src/unstable/library/network/openssl/src/crypto/asymetric/algoritms/ssl_rsa.e:298]: I see, just to be uniform with the rest of the code.

File: Src/unstable/library/network/openssl/src/crypto/ciphers/wrapper/ssl_crypto_externals.e (hunk 20 above)
Comment (jvelilla) [reply Src/unstable/library/network/openssl/src/crypto/ciphers/wrapper/ssl_crypto_externals.e:469]: yes, I was lazy and I was not sure if add it to the feature level or at class level

File: Src/unstable/library/network/openssl/test/application.e (hunk 22 above)
Comment (jvelilla) [reply Src/unstable/library/network/openssl/test/application.e:21]: That's something I need to check for Crypto it seems OpenSSL initialize all the algorithms and errors by default.

**Comment by jocelyn** [comment]
Already merged into eiffelstudio svn repository.

**Pull request closed without merging**